les/lespay/server/balance_tracker.go

```diff
@@ -0,0 +1,297 @@
+	}
+}
+
+// Connect sets the fields used by BalanceTracker as an input
+func (bts *BalanceTrackerSetup) Connect(connAddressField, capacityField nodestate.Field) {
+	bts.connAddressField = connAddressField
+	bts.capacityField = capacityField
+}
+
+// BalanceTracker tracks positive and negative balances for connected nodes.
+// After connAddressField is set externally, a NodeBalance is created and previous
+// balance values are loaded from the database. Both balances are exponentially expired
+// values. Costs are deducted from the positive balance if present, otherwise added to
+// the negative balance. If the capacity is non-zero then a time cost is applied
+// continuously while individual request costs are applied immediately.
+// The two balances are translated into a single priority value that also depends
+// on the actual capacity.
+type BalanceTracker struct {
+	BalanceTrackerSetup
+	clock              mclock.Clock
+	lock               sync.Mutex
+	ns                 *nodestate.NodeStateMachine
+	ndb                *nodeDB
+	posExp, negExp     utils.ValueExpirer
+	posExpTC, negExpTC uint64
+
+	active, inactive utils.ExpiredValue
+	balanceTimer     *utils.UpdateTimer
+	quit             chan struct{}
+}
+
+// NewBalanceTracker creates a new BalanceTracker
+func NewBalanceTracker(ns *nodestate.NodeStateMachine, setup BalanceTrackerSetup, db ethdb.KeyValueStore, clock mclock.Clock, posExp, negExp utils.ValueExpirer) *BalanceTracker {
+	ndb := newNodeDB(db, clock)
+	bt := &BalanceTracker{
+		ns:                  ns,
+		BalanceTrackerSetup: setup,
+		ndb:                 ndb,
+		clock:               clock,
+		posExp:              posExp,
+		negExp:              negExp,
+		balanceTimer:        utils.NewUpdateTimer(clock, time.Second*10),
+		quit:                make(chan struct{}),
+	}
+	bt.ndb.forEachBalance(false, func(id enode.ID, balance utils.ExpiredValue) bool {
+		bt.inactive.AddExp(balance)
+		return true
+	})
+
+	ns.SubscribeField(bt.capacityField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		n, _ := ns.GetField(node, bt.BalanceField).(*NodeBalance)
+		if n == nil {
+			log.Error("Capacity field changed while node field is missing")
+			return
+		}
+
+		ov, _ := oldValue.(uint64)
+		nv, _ := newValue.(uint64)
+		if ov == 0 && nv != 0 {
+			n.activate()
+		}
+		if nv != 0 {
+			n.setCapacity(nv)
+		}
+		if ov != 0 && nv == 0 {
+			n.deactivate()
+		}
+	})
+
+	ns.SubscribeField(bt.BalanceField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if newValue == nil {
+			oldValue.(*NodeBalance).deactivate()
+		}
+	})
+
+	ns.SubscribeField(bt.connAddressField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if newValue != nil {
+			ns.SetFieldSub(node, bt.BalanceField, bt.newNodeBalance(node, newValue.(string)))
+		} else {
+			ns.SetStateSub(node, nodestate.Flags{}, bt.PriorityFlag, 0)
+			ns.SetFieldSub(node, bt.BalanceField, nil)
```

Comment:
I would really like that we move the deactivate logic here.

So that it means this callback is the only place which can initialize the BalanceField or set it to nil.

Reply:
done